File: app/src/main/java/com/zacharee1/rctdremoverforlg/misc/SuUtils.java
package com.zacharee1.rctdremoverforlg.misc;

import android.app.Activity;
import android.content.Context;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import android.widget.ScrollView;
import android.widget.TextView;

import org.zeroturnaround.exec.ProcessExecutor;
import org.zeroturnaround.exec.stream.LogOutputStream;

import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SuUtils
{
    public static Process sudo(String... strings) throws IOException {
        Process su = Runtime.getRuntime().exec("su");

        try{
            DataOutputStream outputStream = new DataOutputStream(su.getOutputStream());

            BufferedReader inputReader = new BufferedReader(new InputStreamReader(su.getInputStream()));
            BufferedReader errorReader = new BufferedReader(new InputStreamReader(su.getErrorStream()));

            for (String s : strings) {
                Log.e("Sudo Exec", s);
                outputStream.writeBytes(s + "\n");
                outputStream.flush();

                String line;

//                while ((line = inputReader.readLine()) != null) {
//                    Log.e("Sudo", line);
//                }
//
//                while ((line = errorReader.readLine()) != null) {
//                    Log.e("Sudo", line);
//                }
            }

            outputStream.writeBytes("exit\n");
            outputStream.flush();
            try {
                su.waitFor();
                Log.e("Done", "Done");
            } catch (InterruptedException e) {
                e.printStackTrace();
                Log.e("No Root?", e.getMessage());
            }
            outputStream.close();
        } catch (IOException e){
            e.printStackTrace();
            throw new IOException(e);
        } catch (Exception e) {
            e.printStackTrace();
        }

        return su;
    }

    public static String sudoForResult(String...strings) {
        String res = "";
        DataOutputStream outputStream = null;
        InputStream response = null;
        try{
            Process su = Runtime.getRuntime().exec("su");
            outputStream = new DataOutputStream(su.getOutputStream());
            response = su.getInputStream();

            for (String s : strings) {
                outputStream.writeBytes(s+"\n");
                outputStream.flush();
            }

            outputStream.writeBytes("exit\n");
            outputStream.flush();

            res = readFully(response);

            outputStream.close();
        } catch (Exception e){
            e.printStackTrace();
        }
        return res;
    }

    public static String readFully(InputStream is) throws IOException {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        byte[] buffer = new byte[1024];
        int length = 0;

        while ((length = is.read(buffer)) > 5) {
            baos.write(buffer, 0, length);

            if (length == 6) break;
        }

        return baos.toString("UTF-8");
    }

    public static String suAndPrintToView(final ProcessExecutor processExecutor, final Activity activity, final TerminalView addTo, ArrayList<String> commands) {
        final StringBuilder builder = new StringBuilder();

        try {
            processExecutor.command("su", "-c", TextUtils.join(" ; ", commands))
                    .redirectOutput(new LogOutputStream() {
                        @Override
                        protected void processLine(final String line) {
                            activity.runOnUiThread(new Runnable() {
                                @Override
                                public void run() {
                                    addTo.addText(line + "\n");
                                    builder.append(line);
                                    final ViewParent scroll = (addTo.getParent()).getParent();

                                    if (scroll instanceof ScrollView) {
                                        ((ScrollView) scroll).post(new Runnable() {
                                            @Override
                                            public void run() {
                                                ((ScrollView) scroll).fullScroll(ScrollView.FOCUS_DOWN);
                                            }
                                        });
                                    }
                                }
                            });
                        }
                    })
                    .execute();
        } catch (Exception e) {
            e.printStackTrace();
        }

        return builder.toString();
    }

    public static boolean testSudo() {
        StackTraceElement st = null;

        try{
            Process su = Runtime.getRuntime().exec("su");
            DataOutputStream outputStream = new DataOutputStream(su.getOutputStream());

            outputStream.writeBytes("exit\n");
            outputStream.flush();

            DataInputStream inputStream = new DataInputStream(su.getInputStream());
            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(inputStream));

            while (bufferedReader.readLine() != null) {
                bufferedReader.readLine();
            }

            su.waitFor();
        } catch (Exception e) {
            e.printStackTrace();
            for (StackTraceElement s : e.getStackTrace()) {
                st = s;
                if (st != null) break;
            }
        }

        return st == null;
    }
}

File: app/src/main/java/com/zacharee1/rctdremoverforlg/misc/SuUtils.kt
package com.zacharee1.rctdremoverforlg.misc

import android.app.Activity
import android.text.TextUtils
import android.util.Log
import android.widget.ScrollView
import org.zeroturnaround.exec.ProcessExecutor
import org.zeroturnaround.exec.ProcessResult
import org.zeroturnaround.exec.stream.LogOutputStream
import java.io.ByteArrayOutputStream
import java.io.DataOutputStream
import java.io.IOException
import java.io.InputStream
import java.util.ArrayList

class SuUtils {
    companion object {
        fun sudo(vararg strings: String): Process {
            val su = Runtime.getRuntime().exec("su")

            try {
                val outputStream = DataOutputStream(su.outputStream)

                for (s in strings) {
                    Log.e("Sudo Exec", s)
                    outputStream.writeBytes(s + "\n")
                    outputStream.flush()
                }

                outputStream.writeBytes("exit\n")
                outputStream.flush()
                try {
                    su.waitFor()
                    Log.e("Done", "Done")
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                    Log.e("No Root?", e.message)
                }

                outputStream.close()
            } catch (e: IOException) {
                e.printStackTrace()
                throw IOException(e)
            } catch (e: Exception) {
                e.printStackTrace()
            }

            return su
        }

        fun sudoForResult(vararg strings: String): String {
            var res = ""
            var outputStream: DataOutputStream? = null
            var response: InputStream? = null
            try {
                val su = Runtime.getRuntime().exec("su")
                outputStream = DataOutputStream(su.outputStream)
                response = su.inputStream

                for (s in strings) {
                    outputStream.writeBytes(s + "\n")
                    outputStream.flush()
                }

                outputStream.writeBytes("exit\n")
                outputStream.flush()

                res = readFully(response!!)

                outputStream.close()
            } catch (e: Exception) {
                e.printStackTrace()
            }

            return res
        }

        fun readFully(inputStream: InputStream): String {
            val baos = ByteArrayOutputStream()
            val buffer = ByteArray(1024)
            var length: Int

            do {
                length = inputStream.read(buffer)

                if (length < 5) break
                baos.write(buffer, 0, length)
                if (length == 6) break
            } while (true)

            return baos.toString("UTF-8")
        }

        fun suAndPrintToView(processExecutor: ProcessExecutor, activity: Activity?, addTo: TerminalView, commands: ArrayList<String>): ProcessResult? {
            val origCmds = ArrayList(commands)

            for (i in origCmds.indices) {
                val cmd = origCmds[i]
                if (!cmd.startsWith("echo")) {
                    val echo = "echo '<font color=\"#ffff00\">" + cmd + "</font>'".replace("|| exit 1", "")
                    commands.add(i, echo)
                }
            }

            try {
                return processExecutor.command("su", "-c", TextUtils.join(" ; ", commands))
                        .readOutput(true)
                        .redirectOutput(object : LogOutputStream() {
                            override fun processLine(line: String) {
                                activity?.runOnUiThread {
                                    addTo.addText(line + "\n")
                                    val scroll = addTo.parent.parent

                                    (scroll as? ScrollView)?.post { scroll.fullScroll(ScrollView.FOCUS_DOWN) }
                                }
                            }
                        })
                        .execute()
            } catch (e: Exception) {
                e.printStackTrace()
            }

            return null
        }

        fun testSudo(): Boolean {
            val commandResult: String

            try {
                commandResult = ProcessExecutor("su")
                        .readOutput(true)
                        .execute()
                        .outputString()
            } catch (e: Exception) {
                return false
            }

            return commandResult.isEmpty()
        }
    }
}